indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskRunnerTest.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.overlord;
+
+import com.google.common.base.Optional;
+import com.google.common.util.concurrent.ListenableFuture;
+import org.apache.druid.indexer.TaskStatus;
+import org.apache.druid.indexing.common.task.Task;
+import org.apache.druid.indexing.overlord.autoscaling.ScalingStats;
+import org.apache.druid.java.util.common.Pair;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.easymock.EasyMock;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class TaskRunnerTest
+{
+  @Test
+  public void testShutdownWithInfoOnConstructsString()
+  {
+    final ToStringMock expensiveObject = new ToStringMock();
+
+    Logger infoLogger = EasyMock.createMock(Logger.class);
+    EasyMock.expect(infoLogger.isDebugEnabled()).andReturn(true).atLeastOnce();
+    EasyMock.replay(infoLogger);
+
+    MockTaskRunner taskRunner = new MockTaskRunner(infoLogger);
+    taskRunner.shutdown("taskid", "reason %s", new Object[]{expensiveObject});
+
+    Assert.assertTrue(expensiveObject.getWasCalled());
+  }
+
+  @Test
+  public void testShutdownWithInfoOffDoesNotCallToString()
+  {
+    final ToStringMock expensiveObject = new ToStringMock();
+
+    Logger infoLogger = EasyMock.createMock(Logger.class);
+    EasyMock.expect(infoLogger.isDebugEnabled()).andReturn(false).atLeastOnce();
+    EasyMock.replay(infoLogger);
+
+    MockTaskRunner taskRunner = new MockTaskRunner(infoLogger);
+    taskRunner.shutdown("taskid", "reason %s", new Object[]{expensiveObject});
+
+    Assert.assertFalse(expensiveObject.getWasCalled());
+  }
+
+  /**
+   * EasyMock does not support mocking of toString, so this provides a custom
+   * object implementation to track whether toString was called.
+   */
+  public static class ToStringMock
```

Comment:
nit: We should probably have these inner classes `ToStringMock` and `MockTaskRunner` as private static.